tests/unittest/test_sampled_logits.py

```diff
@@ -26,51 +26,57 @@
+        batch_size = 2
+        num_sampled = 3
+        vocab_size = 10
+        num_hidden = 5
+        model = f(vocab_size, num_sampled, num_hidden)
+        loss = gluon.loss.SigmoidBCELoss()
+        model.hybridize()
+        model.initialize(mx.init.Xavier(), ctx=ctx)
+        trainer = mx.gluon.Trainer(model.collect_params(), 'sgd')
+        x = mx.nd.ones((batch_size, num_hidden))
+        y = mx.nd.ones((batch_size,))
+        sampled_cls = mx.nd.ones((num_sampled,))
+        sampled_cls_cnt = mx.nd.ones((num_sampled,))
+        true_cls_cnt = mx.nd.ones((batch_size,))
+        samples = (sampled_cls, sampled_cls_cnt, true_cls_cnt)
+        with mx.autograd.record():
+            pred, new_y = model(x, samples, y)
+            assert pred.shape == (batch_size, 1+num_sampled)
+            assert new_y.shape == (batch_size, 1+num_sampled)
+            l = loss(pred, new_y)
+        l.backward()
+        mx.nd.waitall()
+    check_nce_loss(nlp.model.NCEDense)
+    check_nce_loss(nlp.model.SparseNCEDense)
 
 def test_is_softmax_loss():
-    ctx = mx.cpu()
-    batch_size = 2
-    num_sampled = 3
-    vocab_size = 10
-    num_hidden = 5
-    model = nlp.model.ISDense(vocab_size, num_sampled, num_hidden)
-    loss = gluon.loss.SoftmaxCrossEntropyLoss()
-    model.hybridize()
-    model.initialize(mx.init.Xavier(), ctx=ctx)
-    trainer = mx.gluon.Trainer(model.collect_params(), 'sgd')
-    x = mx.nd.ones((batch_size, num_hidden))
-    y = mx.nd.ones((batch_size,))
-    sampled_cls = mx.nd.ones((num_sampled,))
-    sampled_cls_cnt = mx.nd.ones((num_sampled,))
-    true_cls_cnt = mx.nd.ones((batch_size,))
-    samples = (sampled_cls, sampled_cls_cnt, true_cls_cnt)
-    with mx.autograd.record():
-        pred, new_y = model(x, samples, y)
-        assert pred.shape == (batch_size, 1+num_sampled)
-        assert new_y.shape == (batch_size,)
-        l = loss(pred, new_y)
-    l.backward()
-    mx.nd.waitall()
+    def check_is_softmax_loss(f):
+        ctx = mx.cpu()
+        batch_size = 2
+        num_sampled = 3
+        vocab_size = 10
+        num_hidden = 5
+        model = f(vocab_size, num_sampled, num_hidden)
+        loss = gluon.loss.SoftmaxCrossEntropyLoss()
+        model.hybridize()
+        model.initialize(mx.init.Xavier(), ctx=ctx)
+        trainer = mx.gluon.Trainer(model.collect_params(), 'sgd')
+        x = mx.nd.ones((batch_size, num_hidden))
+        y = mx.nd.ones((batch_size,))
+        sampled_cls = mx.nd.ones((num_sampled,))
+        sampled_cls_cnt = mx.nd.ones((num_sampled,))
+        true_cls_cnt = mx.nd.ones((batch_size,))
+        samples = (sampled_cls, sampled_cls_cnt, true_cls_cnt)
+        with mx.autograd.record():
+            pred, new_y = model(x, samples, y)
+            assert pred.shape == (batch_size, 1+num_sampled)
+            assert new_y.shape == (batch_size,)
+            l = loss(pred, new_y)
+        l.backward()
+        mx.nd.waitall()
+    check_is_softmax_loss(nlp.model.ISDense)
```

Comment:
while this is perfectly reasonable, the same can be done with pytest.mark.parameterize, FYI. see https://github.com/dmlc/gluon-nlp/blob/master/tests/unittest/test_sequence_sampler.py#L45-L48

Reply:
I'll do it that way next time